---
title: "synthetic_genome-analysis"
author: "Francine Camacho"
date: "7/31/2019"
output:
  pdf_document: default
  html_document:
    smart: no
---

```{r, message = FALSE }
# load neccessary packages 
require(tidyverse) 
require(ggsci)
require(ggpubr)
require(reticulate)
require(config)
```

### Load configuration. EDIT CONFIG FILE HERE. 
```{r setup, include=FALSE}
exeDir <- getwd()
config <- config::get(file = file.path(exeDir,"SiderophoresConfig.yaml"))
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = config$DataRoot)
```
### EDITING BEYOND THIS POINT FOR ADVANCED USERS ONLY...

### Perform segmentation and build HMMs
```{r, message = FALSE }
dir.create("output", showWarnings = FALSE)
dir.create(config$OutputFiles.HMMOutDir, showWarnings = FALSE)
dir.create(config$OutputFiles.HMMHighPerfOutDir, showWarnings = FALSE)
dir.create(config$OutputFiles.SampleReadIdDir, showWarnings = FALSE)
source_python(file.path(exeDir,"CreateSpHMMs.py"))
start <- strtoi(config$InputParam.Seq_start, 32L)
end <- strtoi(config$InputParam.Seq_end, 32L)
GenerateSpHMM(config$InputFiles.ProtAlnFasta,10L,30L,file.path(getwd(),config$OutputFiles.HMMOutDir),config$InputParam.HMM_Model_Name,start,end)
```

### Load segmented profiled lantibiotics HMMs for synthetic genomes
```{r, message = FALSE}
#load lantibiotics HMM data and recode sampleType for the complexity of the synthetic sample (#of genomes in samples)
lantibiotics <- read_delim(config$InputFiles.HMMRun,col_names = F, delim = "\t") 
df_hmm_cutoff_scores <- read_tsv(config$InputFiles.HMM_Cutoff_Scores, col_names = T)

names(lantibiotics) <- c("readID", "sampleType", "sampleID", "cyclaseType", "HMMScore", "window","interval")

lantibiotics$sampleType[str_detect(lantibiotics$sampleID, "high") ==TRUE] <-"high"
lantibiotics$sampleType[str_detect(lantibiotics$sampleID, "low") ==TRUE] <-"low"

all_intervals <- unique(lantibiotics$interval)
scored_intervals <- unique(df_hmm_cutoff_scores$Interval)
ignored_intervals <- setdiff(all_intervals, scored_intervals)
lantibiotics$interval <- factor(lantibiotics$interval,levels = df_hmm_cutoff_scores$Interval)

# because of gaps in the HMM we will ignore other intervals 
lantibiotics_df <- lantibiotics %>% filter(!interval %in% ignored_intervals )


```
# remove Tp+other genomes that antiSMASH failed to identify 
```{r}
# We need to remove reads that are from TP+Other genomes that were missed by antiSMASH 
if(file.exists(config$InputFiles.BLAST_IGNORE)){
  remove_reads <- read_tsv(config$InputFiles.BLAST_IGNORE, col_names =  T) %>% filter(readCheck == "TP" & lanc_genome == "other") %>% select(c(readID, Sample))
} else {
  remove_reads <- data.frame(Sample=character(),
                 readID=character(),
                 stitle=character(),
                 readCheck=character(),
                 lanc_genome=character())
}
```

##### Remove the reading frame information from the readID. 
```{r,message=FALSE}
# Function to aggregate identitical sample reads located at different reading frames 
# and take the frame with the highest HMM score
formatHMM<-function(hmmdf){
  hmmdfRecoded <- separate(hmmdf, readID, into = c("readIDOnly","F_R_read_frame"), sep = "/", extra = "merge")
  hmmdfRecoded_FR <- separate(hmmdfRecoded, F_R_read_frame, into = c("F_R","frameNumb"), sep = "_", extra = "merge")
  hmmdfRecodedDF<- within(hmmdfRecoded_FR, readID <- paste(readIDOnly,F_R, sep='/'))
  hmmdfRecodedDFUnique<-aggregate(HMMScore ~ readID + sampleID + sampleType + cyclaseType + window + interval, hmmdfRecodedDF, max)
  colnames(hmmdfRecodedDFUnique)<-c("readID","Sample", "sampleType", "cyclaseType", "window", "interval","HMMScore")
  return(hmmdfRecodedDFUnique)
}

#Keep duplicated reads if they are in different reads
lantibiotics_recoded<-formatHMM(lantibiotics_df) %>% anti_join(.,remove_reads)

```


##### Load the BLAST data for genes against our synthetic dataset. 
```{r, message=FALSE}
# BLAST unfiltered reads at 95% pident no readCoverage filter
all_lancBlastDF <- read_delim(config$InputFiles.BLAST_TP_NoCov, col_names = T, delim = "\t")%>% anti_join(.,remove_reads, by= c("sseqid"= "readID", "Sample"))

```

##### Positional information about domains and their locations in respect to the 30_10 spHMM models 
```{r, message=FALSE}
#load siderophore interval positions data 
lanc_positions <- read_tsv(config$InputFiles.GeneIntervalPos,col_names = T)  %>% filter(start != 0)
```

#### Positional read analysis in respect to location mapped to siderophore domain 
##### Keep reads that map to the interval of a given model and covers the model 90% 
```{r, message=FALSE}
filter_blast <- function(df, pos_df){
  results <- data.frame()
  for (i in 1:nrow(pos_df)){
    gene_data <- pos_df[i,]
    #print(gene_data$gene_name)
    #filters datatframe for edges and internal reads compared to model interval
    interval_df_1 <- df %>% filter(qseqid ==gene_data$gene_name) %>%
      filter(qstart %in% gene_data$start:gene_data$end | qend %in% gene_data$start:gene_data$end)
    # need to get reads that are bigger than the interval 
    interval_df_2 <- df %>% filter(qseqid ==gene_data$gene_name) %>% 
      filter(qstart < gene_data$start & qend > gene_data$end)
    interval_df <- rbind(interval_df_1,interval_df_2)
    if (nrow(interval_df) > 0){
      for (j in 1:nrow(interval_df)){
        in_interval_count <- sum(interval_df[j,]$qstart:interval_df[j,]$qend %in% gene_data$start:gene_data$end)
        interval_count <- length(gene_data$start:gene_data$end)
        interval_cov <- (in_interval_count/interval_count) * 100
        if (interval_cov >=90){
          res_df<-interval_df[j,]
          res_df$model_cov <- interval_cov
          res_df$interval <- gene_data$interval
          results <- rbind(results,res_df)
        }
      }
      
    }
    
  }
  return(results)
}
```

```{r, message=FALSE}
#Filter BLAST reads that are within the lantibiotics genes intervals and cover 90% of the model interval
if(file.exists(config$InputFiles.GeneIntervalBlast)){
  lanc_blast_intervals<- read_tsv(config$InputFiles.GeneIntervalBlast, col_names = T) %>% anti_join(.,remove_reads, by= c("sseqid"= "readID", "Sample"))  
} else {
  lanc_blast_intervals<- filter_blast(all_lancBlastDF, lanc_positions)
}
```

##### Determine manual cutoffs for each interval within the model

```{r, message=FALSE}
compare_reads <- function(hmm_df, blast_df){
  names(blast_df)[names(blast_df)=="sseqid"] <-"readID"
  names(blast_df)[names(blast_df)=="qseqid"] <-"bgcName"
  names(hmm_df)[names(hmm_df)=="cyclaseType"] <-"cyclase_type"
  
  # remove columns to compare the two dataframe 
  blastDF <- blast_df %>% select(-c(model_cov, interval))
  common_reads <- hmm_df %>% semi_join(.,blastDF) 
  common_reads$readCheck<-"common-read"
  hmm_unique_reads <- hmm_df %>% anti_join(.,blastDF)
  hmm_unique_reads$readCheck<-"hmm-unique-read"
  compared_data <- rbind(common_reads, hmm_unique_reads)
  return(compared_data)
}

lanc_bin <- compare_reads(lantibiotics_recoded, lanc_blast_intervals)
names(lanc_bin)[names(lanc_bin)=="cyclase_type"] <-"cyclaseType"
```

##### Trying to determine the HMM cutoffs; calculate median 
```{r, message=FALSE}
lanc_bin %>% filter(readCheck == "common-read") %>% group_by(interval) %>% mutate(medianScore = round(median(HMMScore))) %>% distinct(interval, readCheck, medianScore) %>% ungroup()
```

##### Remove HMM unique reads that mapped to BLAST but didn't make the 90% interval model coverage cutoff imposed. These reads shouldn't be considered hmm-unique
```{r, message=FALSE}
#Function to find the hmm_unique reads that are also mapped using BLAST but didn't make 90% interval
compare_hmm_unique <- function(hmm_df, blast_df, pos_df){

  #hmm_unique_df <- hmm_df  %>% inner_join(.,blast_df, by = c("readID"="sseqid", "Sample", "sampleType","cyclaseType"= "cyclase_type"))
   hmm_unique_df <- hmm_df  %>% inner_join(.,blast_df, by = c("readID"="sseqid", "Sample", "sampleType", "cyclase_type"))
  intervals <- unique(hmm_unique_df$interval)
  results <- data.frame()
  #check that reads are in the same interval to throw out 
  for (i in 1:length(intervals)){
    gene_interval_data <- pos_df %>% filter(interval == intervals[i])
    for (k in 1:nrow(gene_interval_data)){
        gene_data <- gene_interval_data[k,]
        #filters datatframe for edges and internal reads compared to model interval
        interval_df  <- hmm_unique_df %>% filter(qseqid ==gene_data$gene_name) %>%
          filter(qstart %in% gene_data$start:gene_data$end | qend %in% gene_data$start:gene_data$end)
        if (nrow(interval_df) > 0 ){
           results <- rbind(results,interval_df)
        }
    
  }
 }  
  return(results%>% distinct())
}
```

### Apply different HMM Scores cutoff to the HMM dataset
```{r,message=FALSE}
#Filter lantibiotics data with cutoffs to compare to BLAST interval reads 

lanc_filtered_median <- lantibiotics_recoded
lanc_filtered_median <- lanc_filtered_median[0,]
for (i in 1:nrow(df_hmm_cutoff_scores)) {
  intervalStr <- toString(df_hmm_cutoff_scores[i, "Interval"])
  cutoffScore <- as.numeric(df_hmm_cutoff_scores[i, "Cutoff"])
  lanc_filtered_median <- rbind(lanc_filtered_median,lantibiotics_recoded %>% filter((interval == intervalStr & HMMScore>=cutoffScore)))
}

lanc_filtered_median_subfive <- lantibiotics_recoded
lanc_filtered_median_subfive <- lanc_filtered_median_subfive[0,]
for (i in 1:nrow(df_hmm_cutoff_scores)) {
  intervalStr <- toString(df_hmm_cutoff_scores[i, "Interval"])
  cutoffScore <- as.numeric(df_hmm_cutoff_scores[i, "Cutoff"]) - 5.0
  lanc_filtered_median_subfive <- rbind(lanc_filtered_median_subfive,lantibiotics_recoded %>% filter((interval == intervalStr & HMMScore>=cutoffScore)))
}

lanc_filtered_median_plusfive <- lantibiotics_recoded
lanc_filtered_median_plusfive <- lanc_filtered_median_plusfive[0,]
for (i in 1:nrow(df_hmm_cutoff_scores)) {
  intervalStr <- toString(df_hmm_cutoff_scores[i, "Interval"])
  cutoffScore <- as.numeric(df_hmm_cutoff_scores[i, "Cutoff"]) + 5.0
  lanc_filtered_median_plusfive <- rbind(lanc_filtered_median_plusfive,lantibiotics_recoded %>% filter((interval == intervalStr & HMMScore>=cutoffScore)))
}

```

```{r, message=FALSE}
#return the hmm-unique reads 
if(file.exists(config$InputFiles.ScaffoldCheck)){
  genomes_check<- read_tsv(config$InputFiles.ScaffoldCheck, col_names = T)
} else {
  genomes_check <- data.frame(Type=character(),ScaffoldList=character())
}

return_hmm_unique <- function(hmm_df, blast_df){
  oxyn_hmm_df <- hmm_df  %>% select(-c(window))
  oxyn_hmm_df$interval <- as.character(oxyn_hmm_df$interval)
  names(oxyn_hmm_df)[names(oxyn_hmm_df)=="cyclaseType"] <-"cyclase_type"
  names(blast_df)[names(blast_df)=="sseqid"] <-"readID"
  OxyN_hmm_unique <- oxyn_hmm_df %>% anti_join(.,blast_df, by= c("readID", "Sample", "sampleType", "interval"))
  return(OxyN_hmm_unique)
}

lanc_positions$cyclase_type <- config$InputParam.HMM_Model_Name
all_lancBlastDF$cyclase_type <- config$InputParam.HMM_Model_Name
median_lanc_hmmunique <-return_hmm_unique(lanc_filtered_median, lanc_blast_intervals)

median_lanc_hmmunique_less_model_cov<-compare_hmm_unique(median_lanc_hmmunique,all_lancBlastDF,lanc_positions ) 

median_remaining_hmm_lanc<- median_lanc_hmmunique %>% anti_join(.,median_lanc_hmmunique_less_model_cov)
table(median_remaining_hmm_lanc$interval)

genomes_median_check <- toString(genomes_check[which(genomes_check$Type=="median_check"),"ScaffoldList"])
genomes_median_check <- unlist(strsplit(genomes_median_check,","))

if(length(genomes_median_check) > 0) {
  remaining_hmm_lanc_pos_median_check <- median_remaining_hmm_lanc %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% filter(scaffold_id %in% genomes_median_check)
} else {
    remaining_hmm_lanc_pos_median_check <- median_remaining_hmm_lanc %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F)
}
  
############################################################################################################################
subfive_lanc_hmmunique <-return_hmm_unique(lanc_filtered_median_subfive, lanc_blast_intervals)

subfive_lanc_hmmunique_less_model_cov<-compare_hmm_unique(subfive_lanc_hmmunique,all_lancBlastDF,lanc_positions ) 

subfive_remaining_hmm_lanc<- subfive_lanc_hmmunique %>% anti_join(.,subfive_lanc_hmmunique_less_model_cov)
table(subfive_remaining_hmm_lanc$interval)

genomes_subfive_check <- toString(genomes_check[which(genomes_check$Type=="subfive_check"),"ScaffoldList"])
genomes_subfive_check <- unlist(strsplit(genomes_subfive_check,","))

if(length(genomes_subfive_check) > 0) {
  remaining_hmm_lanc_pos_subfive_check <- subfive_remaining_hmm_lanc %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% filter(scaffold_id %in% genomes_subfive_check)
} else {
  remaining_hmm_lanc_pos_subfive_check <- subfive_remaining_hmm_lanc %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F)
}

#############################################################################################################################

plusfive_lanc_hmmunique <-return_hmm_unique(lanc_filtered_median_plusfive, lanc_blast_intervals)
plusfive_lanc_hmmunique_less_model_cov<-compare_hmm_unique(plusfive_lanc_hmmunique,all_lancBlastDF,lanc_positions ) 

plusfive_remaining_hmm_lanc<- plusfive_lanc_hmmunique %>% anti_join(.,plusfive_lanc_hmmunique_less_model_cov)
table(plusfive_remaining_hmm_lanc$interval)

genomes_plusfive_check <- toString(genomes_check[which(genomes_check$Type=="plusfive_check"),"ScaffoldList"])
genomes_plusfive_check <- unlist(strsplit(genomes_plusfive_check,","))

if(length(genomes_plusfive_check) > 0) {
  remaining_hmm_lanc_pos_plusfive_check <- plusfive_remaining_hmm_lanc %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% filter(scaffold_id %in% genomes_plusfive_check)
} else {
  remaining_hmm_lanc_pos_plusfive_check <- plusfive_remaining_hmm_lanc %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F)
}

```

```{r}
#Function to parse out HMM unique reads for each sample. 
parseReads <- function(HMMdf, modelName, dirname){
  setwd(dirname)
  samples<-unique(HMMdf$Sample)
  for (s in 1:length(samples)){
    currentSample<-samples[s]
    currentSampleResults<- HMMdf %>% filter(Sample == currentSample)
    currentSampleReads<- unique(currentSampleResults$readID)
    fileName<-paste0(currentSample,paste("-",modelName,"-hmm_unique",sep =""), ".txt", sep ="")
    write.table(currentSampleReads,fileName, quote = F, row.names = F, col.names = F )
  }
}
```

```{r}

if(file.exists(config$InputFiles.HMMUniqueBlast)){
  
  blast_hmm_unique <- read_tsv(config$InputFiles.HMMUniqueBlast, col_names = T) %>% separate(qseqid, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% filter(str_detect(sseqid, fixed(scaffold_id))) %>% select(-c(scaffold_id, rest_id))  %>% arrange(evalue) %>% group_by(qseqid, Sample, cutoff) %>% top_n(1) %>% ungroup()
  
} else {
  blast_hmm_unique <- data.frame(sseqid=character(),
                      slen=character(),
                      sstart=character(),
                      send=character(),
                      qseqid=character(),
                      qlen=character(),
                      qstart=character(),
                      qend=character(),
                      pident=character(),
                      evalue=character(),
                      Sample=character(),
                      cutoff=character())
  blast_hmm_unique[-1] <- lapply(blast_hmm_unique[-1], as.character)
}


filter_hmm_unique  <- function(df, pos_df){
  results <- data.frame()
  if(nrow(df)>0){
    for (i in 1:nrow(df)){
      blast_data <- df[i,]
      domain <- blast_data$sseqid
      domain_interval <- blast_data$interval 
      read_start <-blast_data$sstart
      read_end <-blast_data$send
      min_start <- min(read_start, read_end)
      max_end <- max(read_start, read_end)
      gene_data <- pos_df %>% filter( gene_name == domain  & interval == domain_interval)
      #filters dataframe for edges and internal reads compared to model interval
      interval_df_1 <- blast_data %>% filter(sseqid ==gene_data$gene_name) %>%
        filter(min_start %in% gene_data$start:gene_data$end | max_end %in% gene_data$start:gene_data$end)
      # need to get reads that are bigger than the interval 
      interval_df_2 <- df %>% filter(sseqid ==gene_data$gene_name) %>% 
        filter(min_start < gene_data$start & max_end > gene_data$end)
      
      results <- rbind(results,interval_df_1,interval_df_2)
    }
  }
  return(results)
}

median_blast_df <- blast_hmm_unique %>% filter(cutoff== "median_score") %>% inner_join(., remaining_hmm_lanc_pos_median_check, by = c("qseqid" = "readID", "Sample"))

subfive_blast_df <- blast_hmm_unique %>% filter(cutoff== "minus_five_score") %>% inner_join(., remaining_hmm_lanc_pos_subfive_check, by = c("qseqid" = "readID", "Sample"))

plusfive_blast_df <- blast_hmm_unique %>% filter(cutoff== "plus_five_score") %>% inner_join(., remaining_hmm_lanc_pos_plusfive_check, by = c("qseqid" = "readID", "Sample"))


remove_lanc_reads_median <- filter_hmm_unique(median_blast_df,lanc_positions) %>% as.data.frame() %>% distinct(qseqid,Sample) 
if (dim(remove_lanc_reads_median)[2] > 0) {
  names(remove_lanc_reads_median)[1]<-"readID"
} else {
  remove_lanc_reads_median <- data.frame(readID=character(),
                      Sample=character())
  remove_lanc_reads_median[1] <- lapply(remove_lanc_reads_median[1], as.character)
  remove_lanc_reads_median[2] <- lapply(remove_lanc_reads_median[2], as.character)
}
remove_lanc_reads_subfive <- filter_hmm_unique(subfive_blast_df,lanc_positions) %>% as.data.frame() %>% distinct(qseqid, Sample) 
if (dim(remove_lanc_reads_subfive)[2] > 0) {
  names(remove_lanc_reads_subfive)[1]<-"readID"
} else {
  remove_lanc_reads_subfive <- data.frame(readID=character(),
                      Sample=character())
  remove_lanc_reads_subfive[1] <- lapply(remove_lanc_reads_subfive[1], as.character)
  remove_lanc_reads_subfive[2] <- lapply(remove_lanc_reads_subfive[2], as.character)
}
remove_lanc_reads_plusfive <- filter_hmm_unique(plusfive_blast_df,lanc_positions) %>% as.data.frame()%>% distinct(qseqid, Sample)
if (dim(remove_lanc_reads_plusfive)[2] > 0) {
  names(remove_lanc_reads_plusfive)[1]<-"readID"
} else {
  remove_lanc_reads_plusfive <- data.frame(readID=character(),
                      Sample=character())
  remove_lanc_reads_plusfive[1] <- lapply(remove_lanc_reads_plusfive[1], as.character)
  remove_lanc_reads_plusfive[2] <- lapply(remove_lanc_reads_plusfive[2], as.character)
}
```

```{r}
calculate_F1 <- function(hmm_df, hmm_fp, blast_df, intervals){
  #added this because factor vector 
  hmm_df$interval <- as.character(hmm_df$interval)
  hmm_df <- hmm_df %>% select(-c(window))
  names(hmm_df)[names(hmm_df)=="cyclaseType"] <-"cyclase_type"
  names(blast_df)[names(blast_df)=="sseqid"] <-"readID"
  
  results<-data.frame(interval=character(), F1=numeric())
  for (i in 1:length(intervals)){
    model <- intervals[i]
    model_hmm <- hmm_df %>% filter(interval== model) # filter out model HMM results 
    model_blast <- blast_df %>% filter(interval== model) 
    model_hmm_fp <-hmm_fp %>% filter(interval== model) 
    TP <- model_hmm %>% inner_join(.,model_blast) %>% nrow()
    FP <- model_hmm_fp %>% nrow()
    FN <- model_blast %>% anti_join(.,model_hmm) %>% nrow()
    F1_metric <- (2*TP)/((2*TP)+FP+FN)
    row <- data.frame(cbind(model, F1_metric))
    results<- results %>% add_row(interval = model, F1 = F1_metric)
  }
  return(results)
}
  
subfive_false_postives <- subfive_remaining_hmm_lanc %>% anti_join(.,remove_lanc_reads_subfive)


median_false_postives <- median_remaining_hmm_lanc %>% anti_join(.,remove_lanc_reads_median)


plusfive_false_postives <- plusfive_remaining_hmm_lanc %>% anti_join(.,remove_lanc_reads_plusfive)



lanc_f1_cutoff_median<- calculate_F1(lanc_filtered_median, median_false_postives, lanc_blast_intervals, unique(lantibiotics_recoded$interval))
lanc_f1_cutoff_median$cutoff<- "median"
lanc_f1_cutoff_plusfive<- calculate_F1(lanc_filtered_median_plusfive, plusfive_false_postives, lanc_blast_intervals, unique(lantibiotics_recoded$interval))
lanc_f1_cutoff_plusfive$cutoff<- "+5"
lanc_f1_cutoff_subfive<- calculate_F1(lanc_filtered_median_subfive, subfive_false_postives, lanc_blast_intervals, unique(lantibiotics_recoded$interval))
lanc_f1_cutoff_subfive$cutoff<- "-5"

lanc_f1_df <- rbind(lanc_f1_cutoff_median,lanc_f1_cutoff_plusfive,lanc_f1_cutoff_subfive)

```

#Find the optimal cutoffs per interval using F1 metric
```{r}
lanc_f1_cutoff_df <- lanc_f1_df %>% group_by(interval) %>% top_n(1, F1) %>% ungroup() %>%  filter(F1 >=as.numeric(config$InputParam.F1_Threshold)) %>% arrange(interval) %>% as.data.frame()

colnames(lanc_f1_cutoff_df)[colnames(lanc_f1_cutoff_df)=="cutoff"] <- "cutoff_diff"
colnames(lanc_f1_cutoff_df)[colnames(lanc_f1_cutoff_df)=="interval"] <- "Interval"

var <- ("cutoff_diff")
lanc_f1_cutoff_df[,var] <- sapply(lanc_f1_cutoff_df[,var],function(x) ifelse(x=='median',0,x))
lanc_f1_cutoff_df <- base::merge(x=lanc_f1_cutoff_df,y=df_hmm_cutoff_scores,by="Interval")
lanc_f1_cutoff_df$FinalCutoff <- as.numeric(lanc_f1_cutoff_df$cutoff_diff) + as.numeric(lanc_f1_cutoff_df$Cutoff)

lanc_f1_cutoff_df <- subset(lanc_f1_cutoff_df, select = -c(cutoff_diff, Cutoff))

write_tsv(lanc_f1_cutoff_df,config$OutputFiles.F1_Cutoff,col_names = T)

CopyHPModel <- function(x) {
 spHMMInterval <- x[1]
 spHMMFileName = paste0(config$InputParam.HMM_Model_Name, "__30_10__", spHMMInterval, ".hmm")
 file.copy(file.path(config$OutputFiles.HMMOutDir,spHMMFileName), config$OutputFiles.HMMHighPerfOutDir)
}
apply(lanc_f1_cutoff_df, 1, CopyHPModel)
```


```{r}
supp_fig_lanc_f1 <-ggplot(data = lanc_f1_df, mapping = aes(x = interval, y = F1, group= cutoff, colour=cutoff )) + geom_point()+
    geom_line() + ylim(0,1)+ geom_hline(yintercept=c(as.numeric(config$InputParam.F1_Threshold),0.1051712), linetype="dashed", 
                color = c("red","green"), size=1) + theme_pubclean() +  scale_color_npg(name="HMM Score Cutoff")

ggsave(supp_fig_lanc_f1, file=config$OutputFiles.Plot, device="eps", width = 20, height = 7)
```

```{r}
# Finalized cutoffs with intervals and cutoffs 

lanc_filtered_final <- lantibiotics_recoded
lanc_filtered_final <- lanc_filtered_final[0,]
for (i in 1:nrow(lanc_f1_cutoff_df)) {
  intervalStr <- toString(lanc_f1_cutoff_df[i, "Interval"])
  cutoffGroup <- toString(lanc_f1_cutoff_df[i, "FinalCutoff"])
  intervalIndx <- which(df_hmm_cutoff_scores$Interval == intervalStr)
  cutoffScore <- as.numeric(df_hmm_cutoff_scores[intervalIndx,"Cutoff"])
  if(cutoffGroup=="-5"){
    cutoffScore <- cutoffScore-5
  } else if(cutoffGroup=="+5"){
    cutoffScore <- cutoffScore+5
  }
  lanc_filtered_final <- rbind(lanc_filtered_final,lantibiotics_recoded %>% filter((interval == intervalStr & HMMScore>=cutoffScore)))
}

lanc_filtered_final <- distinct(lanc_filtered_final)

#lanc_filtered_final%>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% distinct(scaffold_id)
write_tsv( lanc_filtered_final %>% select(-c(window, interval))%>% group_by(readID,Sample) %>% arrange(desc(HMMScore)) %>% top_n(1, HMMScore) %>% filter(row_number() == 1) %>% ungroup(), config$OutputFiles.RawModelData, col_names = T)

#rename hmm_unique within the folder to detected_reads

parseReads(lanc_filtered_final, config$InputParam.HMM_Model_Name, config$OutputFiles.SampleReadIdDir)
```


## Output FP/TP for HMM reads detected in final models with final cutoffs/ keep reads duplicated in multiple intervals with the highest HMM Score if, they are tied, then choose the first one. 
```{r}

lanc_filtered_final_distinct <- lanc_filtered_final %>% arrange(desc(HMMScore)) %>% group_by(readID, Sample) %>% top_n(1) %>% 
  filter(row_number()==1) %>% ungroup() %>% select(-c(cyclaseType))

lanc_filtered_final_distinct$interval <-as.character(lanc_filtered_final_distinct$interval)
#names(lanc_filtered_final_distinct)[names(lanc_filtered_final_distinct)=="cyclaseType"] <-"cyclase_type"

results_common<-lanc_filtered_final_distinct %>% inner_join(., lanc_blast_intervals, by = c("readID" = "sseqid", "Sample", "sampleType", "interval")) %>% select(-c(model_cov))
results_common$cyclase_type <- config$InputParam.HMM_Model_Name

results_hmm_unique<-lanc_filtered_final_distinct %>% anti_join(., lanc_blast_intervals, by = c("readID" = "sseqid", "Sample", "sampleType", "interval")) 
results_hmm_unique$cyclase_type <- config$InputParam.HMM_Model_Name

results_hmm_unique_TP<-compare_hmm_unique(results_hmm_unique,all_lancBlastDF,lanc_positions ) 


results_hmm_unique_FP <- results_hmm_unique %>% anti_join(.,results_hmm_unique_TP)


lanc_filtered_final_distinct_TPs <- rbind(results_common, results_hmm_unique_TP) %>% select(-c(7:9,11:15))

combined_lanc_results <- lanc_filtered_final_distinct_TPs %>% full_join(., results_hmm_unique_FP )

write_tsv(combined_lanc_results,config$OutputFiles.ModelData,col_names = T)
```

```{r}
sessionInfo()
```