---
title: "TcmI-synthetic_genome-analysis"
author: "Francine Camacho"
date: "2/21/2019"
output:
  html_document: 
    smart: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/francinecamacho/Google Drive/T2PKS-paper/computational_analysis/synthetic_genomes/HMM-analysis/TcmI/version-2/')

```

```{r, message = FALSE }
# load neccessary packages 
require(tidyverse) 
require(ggsci)
require(ggpubr)
```

### Load segmented profiled cyclase TcmI HMMs for synthetic genomes
```{r, message = FALSE}
#load TcmI HMM data and recode sampleType for the complexity of the synthetic sample (#of genomes in samples)
cyclase_TcmI <- read_delim("combined-cyclase_TcmI-spHMM-synthetic_genomes-results.txt",col_names = F, delim = "\t") 
names(cyclase_TcmI) <- c("readID", "sampleType", "sampleID", "cyclaseType", "HMMScore", "window","interval")

cyclase_TcmI$sampleType[str_detect(cyclase_TcmI$sampleID, "high") ==TRUE] <-"high"
cyclase_TcmI$sampleType[str_detect(cyclase_TcmI$sampleID, "low") ==TRUE] <-"low"

cyclase_TcmI$interval <- factor(cyclase_TcmI$interval,levels = c(
  "0_30","10_40","20_50","30_60","40_70","50_80","60_90","70_100","80_110",
  "90_120","100_130","110_140","120_150","130_160","140_170","150_180"))
```

##### Remove the reading frame information from the readID. 
```{r,message=FALSE}
# Function to aggregate identitical sample reads located at different reading frames 
# and take the frame with the highest HMM score
formatHMM<-function(hmmdf){
  hmmdfRecoded <- separate(hmmdf, readID, into = c("readIDOnly","F_R_read_frame"), sep = "/", extra = "merge")
  hmmdfRecoded_FR <- separate(hmmdfRecoded, F_R_read_frame, into = c("F_R","frameNumb"), sep = "_", extra = "merge")
  hmmdfRecodedDF<- within(hmmdfRecoded_FR, readID <- paste(readIDOnly,F_R, sep='/'))
  hmmdfRecodedDFUnique<-aggregate(HMMScore ~ readID + sampleID + sampleType + cyclaseType + window + interval, hmmdfRecodedDF, max)
  colnames(hmmdfRecodedDFUnique)<-c("readID","Sample", "sampleType", "cyclaseType", "window", "interval","HMMScore")
  return(hmmdfRecodedDFUnique)
}

#Keep duplicated reads if they are in different reads
cyclaseTcmI_recoded<-formatHMM(cyclase_TcmI)
```

##### Load the BLAST data for T2PKS cyclase genes against our synthetic dataset. 
```{r, message=FALSE}
#BLAST True positive reads mapped to the T2PKS cyclases genes 
cyclase_metadata <- read_tsv("../../T2PKS-pos_genomes-cyclase_gene-metadata.txt", col_names = T) %>% filter(used_in_synthetic_metagenome == "Y") %>% select(c(11:9))

# BLAST unfiltered reads at 95% pident no readCoverage filter
all_t2pksBlastDF <- read_delim("../../../T2PKS-TP-reads/t2pksBlastDF_all-results.txt", col_names = T, delim = "\t") %>% inner_join(.,cyclase_metadata, by= c("qseqid" = "nucl_cyclase_target_name") )

```

##### Positional information about OxyN domains and their locations in respect to the OxyN 30_10 spHMM models 
```{r, message=FALSE}
#load cyclases interval positions data 
tcmi_positions <- read_tsv("../../../T2PKS-TP-reads/cyclases_genes-interval_position-data.txt",col_names = T)%>% filter(cyclase_type == "TcmI")
```

#### Positional read analysis in respect to location mapped to cyclase OxyN domain 
##### Keep reads that map to the interval of a given model and covers the model 90% 
```{r, message=FALSE}
filter_blast <- function(df, pos_df){
  results <- data.frame()
  for (i in 1:nrow(pos_df)){
    gene_data <- pos_df[i,]
    #filters datatframe for edges and internal reads compared to model interval
    interval_df_1 <- df %>% filter(qseqid ==gene_data$gene_name) %>%
      filter(qstart %in% gene_data$start:gene_data$end | qend %in% gene_data$start:gene_data$end)
    # need to get reads that are bigger than the interval 
    interval_df_2 <- df %>% filter(qseqid ==gene_data$gene_name) %>% 
      filter(qstart < gene_data$start & qend > gene_data$end)
    interval_df <- rbind(interval_df_1,interval_df_2)
    if (nrow(interval_df) > 0){
      for (j in 1:nrow(interval_df)){
        in_interval_count <- sum(interval_df[j,]$qstart:interval_df[j,]$qend %in% gene_data$start:gene_data$end)
        interval_count <- length(gene_data$start:gene_data$end)
        interval_cov <- (in_interval_count/interval_count) * 100
        if (interval_cov >=90){
          res_df<-interval_df[j,]
          res_df$model_cov <- interval_cov
          res_df$interval <- gene_data$interval
          results <- rbind(results,res_df)
        }
      }
      
    }
    
  }
  return(results)
}
```

```{r, message=FALSE}
#Filter BLAST reads that are within the TcmI genes intervals and cover 90% of the model interval
tcmI_blast_intervals<- filter_blast(all_t2pksBlastDF %>% filter(cyclase_type == "TcmI"), tcmi_positions)
```

##### Determine manual cutoffs for each interval within the model

```{r, message=FALSE}
compare_reads <- function(hmm_df, blast_df){
  names(blast_df)[names(blast_df)=="sseqid"] <-"readID"
  names(blast_df)[names(blast_df)=="qseqid"] <-"bgcName"
  names(hmm_df)[names(hmm_df)=="cyclaseType"] <-"cyclase_type"
  
  # remove columns to compare the two dataframe 
  blastDF <- blast_df %>% select(-c(model_cov, interval))
  common_reads <- hmm_df %>% semi_join(.,blastDF) 
  common_reads$readCheck<-"common-read"
  hmm_unique_reads <- hmm_df %>% anti_join(.,blastDF)
  hmm_unique_reads$readCheck<-"hmm-unique-read"
  compared_data <- rbind(common_reads, hmm_unique_reads)
  return(compared_data)
}

cyclaseTcmI_bin <- compare_reads(cyclaseTcmI_recoded, tcmI_blast_intervals)
names(cyclaseTcmI_bin)[names(cyclaseTcmI_bin)=="cyclase_type"] <-"cyclaseType"
```

##### Trying to determine the HMM cutoffs; calculate median 
```{r, message=FALSE}
cyclaseTcmI_bin %>% filter(readCheck == "common-read") %>% group_by(interval) %>% mutate(medianScore = round(median(HMMScore))) %>% distinct(interval, readCheck, medianScore) %>% ungroup()
```

##### Remove HMM unique reads that mapped to BLAST but didn't make the 90% interval model coverage cutoff imposed. These reads shouldn't be considered hmm-unique
```{r, message=FALSE}
#Function to find the hmm_unique reads that are also mapped using BLAST but didn't make 90% interval
compare_hmm_unique <- function(hmm_df, blast_df, pos_df){

  #hmm_unique_df <- hmm_df  %>% inner_join(.,blast_df, by = c("readID"="sseqid", "Sample", "sampleType","cyclaseType"= "cyclase_type"))
   hmm_unique_df <- hmm_df  %>% inner_join(.,blast_df, by = c("readID"="sseqid", "Sample", "sampleType", "cyclase_type"))
  intervals <- unique(hmm_unique_df$interval)
  results <- data.frame()
  #check that reads are in the same interval to throw out 
  for (i in 1:length(intervals)){
    gene_interval_data <- pos_df %>% filter(interval == intervals[i])
    for (k in 1:nrow(gene_interval_data)){
        gene_data <- gene_interval_data[k,]
        #filters datatframe for edges and internal reads compared to model interval
        interval_df  <- hmm_unique_df %>% filter(qseqid ==gene_data$gene_name) %>%
          filter(qstart %in% gene_data$start:gene_data$end | qend %in% gene_data$start:gene_data$end)
        if (nrow(interval_df) > 0 ){
           results <- rbind(results,interval_df)
        }
    
  }
 }  
  return(results%>% distinct())
}
```

### Apply different HMM Scores cutoff to the HMM dataset
```{r,message=FALSE}
#Filter OxyN data with cutoffs to compare to BLAST interval reads 
TcmI_filtered_median <- cyclaseTcmI_recoded %>% filter((interval == "0_30" & HMMScore>=40)|
                                         (interval == "10_40" & HMMScore >=35)| 
                                         (interval == "20_50" & HMMScore >=35)| 
                                         (interval == "30_60" & HMMScore >=35)| 
                                         (interval == "40_70" & HMMScore >=30 )|
                                         (interval == "50_80" & HMMScore>=30)| 
                                         (interval == "60_90" & HMMScore >=25 ) | 
                                         (interval == "70_100" & HMMScore >=35 )| 
                                         (interval == "80_110" & HMMScore >=45 )| 
                                         (interval == "90_120" & HMMScore >=40 ))

TcmI_filtered_median_subfive<- cyclaseTcmI_recoded %>% filter((interval == "0_30" & HMMScore>=35)|
                                         (interval == "10_40" & HMMScore >=30)| 
                                         (interval == "20_50" & HMMScore >=30)| 
                                         (interval == "30_60" & HMMScore >=30)| 
                                         (interval == "40_70" & HMMScore >=25 )|
                                         (interval == "50_80" & HMMScore>=25)| 
                                         (interval == "60_90" & HMMScore >=20 ) | 
                                         (interval == "70_100" & HMMScore >=30 )| 
                                         (interval == "80_110" & HMMScore >=40 )| 
                                         (interval == "90_120" & HMMScore >=35 ))

TcmI_filtered_median_plusfive <- cyclaseTcmI_recoded %>% filter((interval == "0_30" & HMMScore>=45)|
                                         (interval == "10_40" & HMMScore >=40)| 
                                         (interval == "20_50" & HMMScore >=40)| 
                                         (interval == "30_60" & HMMScore >=40)| 
                                         (interval == "40_70" & HMMScore >=35 )|
                                         (interval == "50_80" & HMMScore>=35)| 
                                         (interval == "60_90" & HMMScore >=30 ) | 
                                         (interval == "70_100" & HMMScore >=40 )| 
                                         (interval == "80_110" & HMMScore >=50 )| 
                                         (interval == "90_120" & HMMScore >=45 ))
```

```{r, message=FALSE}
#return the hmm-unique reads 
return_hmm_unique <- function(hmm_df, blast_df){
  oxyn_hmm_df <- hmm_df  %>% select(-c(window))
  oxyn_hmm_df$interval <- as.character(oxyn_hmm_df$interval)
  names(oxyn_hmm_df)[names(oxyn_hmm_df)=="cyclaseType"] <-"cyclase_type"
  names(blast_df)[names(blast_df)=="sseqid"] <-"readID"
  OxyN_hmm_unique <- oxyn_hmm_df %>% anti_join(.,blast_df, by= c("readID", "Sample", "sampleType", "interval"))
  return(OxyN_hmm_unique)
}

median_tcmI_hmmunique <-return_hmm_unique(TcmI_filtered_median, tcmI_blast_intervals)

median_TcmI_hmmunique_less_model_cov<-compare_hmm_unique(median_tcmI_hmmunique,all_t2pksBlastDF,tcmi_positions ) 

median_remaining_hmm_tcmi<- median_tcmI_hmmunique %>% anti_join(.,median_TcmI_hmmunique_less_model_cov)
table(median_remaining_hmm_tcmi$interval)

pks_genomes_median_check <- c("JOFH01000019.1", "gi|529222834|ref|NC_021985.1|", "gi|640932211|gb|JJOH01000028.1|", "AZXI01000002.1")


#median_remaining_hmm_tcmi %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% distinct(scaffold_id)

remaining_hmm_tcmi_pos_median_check <- median_remaining_hmm_tcmi %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% filter(scaffold_id %in% pks_genomes_median_check)
############################################################################################################################
subfive_tcmI_hmmunique <-return_hmm_unique(TcmI_filtered_median_subfive, tcmI_blast_intervals)

subfive_TcmI_hmmunique_less_model_cov<-compare_hmm_unique(subfive_tcmI_hmmunique,all_t2pksBlastDF,tcmi_positions ) 

subfive_remaining_hmm_tcmi<- subfive_tcmI_hmmunique %>% anti_join(.,subfive_TcmI_hmmunique_less_model_cov)
table(subfive_remaining_hmm_tcmi$interval)

#subfive_remaining_hmm_tcmi %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% distinct(scaffold_id)
pks_genomes_subfive_check <- c("JOFH01000019.1", "gi|529222834|ref|NC_021985.1|", "AZXI01000034.1", "gi|928421279|gb|LGKI01000002.1|", "AZXI01000002.1", "gi|640932211|gb|JJOH01000028.1|")

remaining_hmm_tcmi_pos_subfive_check <- subfive_remaining_hmm_tcmi %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% filter(scaffold_id %in% pks_genomes_subfive_check)

###############################################################################################################################################

plusfive_tcmI_hmmunique <-return_hmm_unique(TcmI_filtered_median_plusfive, tcmI_blast_intervals)

plusfive_TcmI_hmmunique_less_model_cov<-compare_hmm_unique(plusfive_tcmI_hmmunique,all_t2pksBlastDF,tcmi_positions ) 

plusfive_remaining_hmm_tcmi<- plusfive_tcmI_hmmunique %>% anti_join(.,plusfive_TcmI_hmmunique_less_model_cov)
table(plusfive_remaining_hmm_tcmi$interval)

#plusfive_remaining_hmm_tcmi %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% distinct(scaffold_id)

pks_genomes_plusfive_check <- c("gi|529222834|ref|NC_021985.1|", "JOFH01000019.1")

remaining_hmm_tcmi_pos_plusfive_check <- plusfive_remaining_hmm_tcmi %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% filter(scaffold_id %in% pks_genomes_plusfive_check)

```

```{r}
#Function to parse out HMM unique reads for each sample. 
parseReads <- function(HMMdf, modelName, dirname){
  setwd(dirname)
  samples<-unique(HMMdf$Sample)
  for (s in 1:length(samples)){
    currentSample<-samples[s]
    currentSampleResults<- HMMdf %>% filter(Sample == currentSample)
    currentSampleReads<- unique(currentSampleResults$readID)
    fileName<-paste0(currentSample,paste("-",modelName,"-hmm_unique",sep =""), ".txt", sep ="")
    write.table(currentSampleReads,fileName, quote = F, row.names = F, col.names = F )
  }
}

#parseReads(remaining_hmm_tcmi_pos_median_check, "cyclaseTcmI", "/Users/francinecamacho/Google Drive/T2PKS-paper/computational_analysis/synthetic_genomes/HMM-analysis/TcmI/version-2/hmm-unique-analysis/median_score/")

#parseReads(remaining_hmm_tcmi_pos_subfive_check, "cyclaseTcmI", "/Users/francinecamacho/Google Drive/T2PKS-paper/computational_analysis/synthetic_genomes/HMM-analysis/TcmI/version-2/hmm-unique-analysis/minus_five_score/")

#parseReads(remaining_hmm_tcmi_pos_plusfive_check, "cyclaseTcmI", "/Users/francinecamacho/Google Drive/T2PKS-paper/computational_analysis/synthetic_genomes/HMM-analysis/TcmI/version-2/hmm-unique-analysis/plus_five_score/")

```

```{r, message=FALSE}

calculate_F1 <- function(hmm_df, hmm_fp, blast_df, intervals){
  #added this because factor vector 
  hmm_df$interval <- as.character(hmm_df$interval)
  hmm_df <- hmm_df %>% select(-c(window))
  names(hmm_df)[names(hmm_df)=="cyclaseType"] <-"cyclase_type"
  names(blast_df)[names(blast_df)=="sseqid"] <-"readID"
  
  results<-data.frame(interval=character(), F1=numeric())
  for (i in 1:length(intervals)){
    model <- intervals[i]
    model_hmm <- hmm_df %>% filter(interval== model) # filter out model HMM results 
    model_blast <- blast_df %>% filter(interval== model) 
    model_hmm_fp <-hmm_fp %>% filter(interval== model) 
    TP <- model_hmm %>% inner_join(.,model_blast) %>% nrow()
    FP <- model_hmm_fp %>% nrow()
    FN <- model_blast %>% anti_join(.,model_hmm) %>% nrow()
    F1_metric <- (2*TP)/((2*TP)+FP+FN)
    row <- data.frame(cbind(model, F1_metric))
    results<- results %>% add_row(interval = model, F1 = F1_metric)
  }
  

    return(results)
  
}


tcmi_f1_cutoff_median <- calculate_F1(TcmI_filtered_median, median_remaining_hmm_tcmi, tcmI_blast_intervals, unique(cyclaseTcmI_recoded$interval))
tcmi_f1_cutoff_median$cutoff<- "median"
tcmi_f1_cutoff_plusfive<- calculate_F1(TcmI_filtered_median_plusfive, plusfive_remaining_hmm_tcmi, tcmI_blast_intervals, unique(cyclaseTcmI_recoded$interval))
tcmi_f1_cutoff_plusfive$cutoff<- "+5"
tcmi_f1_cutoff_subfive<- calculate_F1(TcmI_filtered_median_subfive, subfive_remaining_hmm_tcmi, tcmI_blast_intervals, unique(cyclaseTcmI_recoded$interval))
tcmi_f1_cutoff_subfive$cutoff<- "-5"

tcmi_f1_df <- rbind(tcmi_f1_cutoff_median,tcmi_f1_cutoff_plusfive,tcmi_f1_cutoff_subfive )

```

### Plot graph to show the F1 metric at different cutoffs 

```{r}
supp_fig_tcmI_f1 <-ggplot(data = tcmi_f1_df, mapping = aes(x = interval, y = F1, group= cutoff, colour=cutoff )) + geom_point()+
    geom_line() + ylim(0,1)+ geom_hline(yintercept=.50, linetype="dashed", 
                color = "red", size=1) + theme_pubclean() +  scale_color_npg(name="HMM Score Cutoff")
#ggsave(supp_fig_tcmI_f1, file="TcmI_synthetic_F1-supp.eps", device="eps", width = 17, height = 7)

```

## Make final TcmI with final cutoffs for the highest F1 for each interval 
```{r}
final_ggplot_tcmi<- tcmi_f1_df %>% group_by(interval) %>% top_n(1, F1) %>% ungroup() %>% ggplot(., mapping = aes(x = interval, y = F1, group= 1 )) + geom_point()+ ylim(0,1)+
    geom_line() +  geom_hline(yintercept=.50, linetype="dashed", 
                color = "red", size=1) + theme_pubclean()+ scale_color_npg()


#ggsave(final_ggplot_tcmi, file="TcmI_synthetic_F1-final.eps", device="eps", width = 17, height = 7)
```

```{r}
# Finalized cutoffs with intervals and cutoffs 
TcmI_filtered_final <- cyclaseTcmI_recoded %>% filter((interval == "0_30" & HMMScore>=35)|
                                         (interval == "10_40" & HMMScore >=30)| 
                                         (interval == "20_50" & HMMScore >=30)| 
                                         (interval == "30_60" & HMMScore >=30)| 
                                         (interval == "40_70" & HMMScore >=25 )|
                                         (interval == "50_80" & HMMScore>=25)| 
                                         (interval == "60_90" & HMMScore >=20 ) | 
                                         (interval == "70_100" & HMMScore >=30 )| 
                                         (interval == "80_110" & HMMScore >=40 )| 
                                         (interval == "90_120" & HMMScore >=35 ))

TcmI_filtered_final_updated <- cyclaseTcmI_recoded %>% filter((interval == "0_30" & HMMScore>=35)|
                                         (interval == "10_40" & HMMScore >=30)| 
                                         (interval == "20_50" & HMMScore >=30)| 
                                         (interval == "30_60" & HMMScore >=30)| 
                                         (interval == "40_70" & HMMScore >=25 )|
                                         (interval == "50_80" & HMMScore>=25)| 
                                         (interval == "70_100" & HMMScore >=30 )| 
                                         (interval == "80_110" & HMMScore >=40 )| 
                                         (interval == "90_120" & HMMScore >=35 ))

#TcmI_filtered_final%>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% distinct(scaffold_id)
#write_tsv( TcmI_filtered_final %>% select(-c(window, interval))%>% group_by(readID,Sample) %>% arrange(desc(HMMScore)) %>% top_n(1, HMMScore) %>% filter(row_number() == 1) %>% ungroup(), "TcmI-final_data-raw.txt", col_names = T)
#rename TcmI hmm_unique within the folder to detected_reads
#parseReads(TcmI_filtered_final, "cyclaseTcmI", "/Users/francinecamacho/Google Drive/T2PKS-paper/computational_analysis/synthetic_genomes/read-quantification/TcmI/sample-readIDs-files")

```

```{r}

final_ggplot_tcmi_full<- tcmi_f1_df %>% group_by(interval) %>% top_n(1, F1) %>% ungroup() %>% ggplot(., mapping = aes(x = interval, y = F1, group= 1 )) + geom_point()+ ylim(0,1)+
    geom_line() +  geom_hline(yintercept=.50, linetype="dashed", 
                color = "red", size=1) + geom_hline(yintercept=0.7641951, linetype="dashed", 
                color = "green", size=1) +theme_pubclean()

#ggsave(final_ggplot_tcmi_full, file="TcmI_synthetic_F1-final-with-Full-F1.eps", device="eps", width = 17, height = 7)
```


## Output FP/TP for HMM reads detected in final models with final cutoffs/ keep reads duplicated in multiple intervals with the highest HMM Score if, they are tied, then choose the first one. 

```{r}
TcmI_filtered_final_distinct <- TcmI_filtered_final %>% arrange(desc(HMMScore)) %>% group_by(readID, Sample) %>% top_n(1) %>% 
  filter(row_number()==1) %>% ungroup() #2473
#TcmI_filtered_final_distinct <- TcmI_filtered_final_updated  %>% arrange(desc(HMMScore)) %>% group_by(readID, Sample) %>% top_n(1) %>% filter(row_number()==1) %>% ungroup() #2473
TcmI_filtered_final_distinct$interval <-as.character(TcmI_filtered_final_distinct$interval)
names(TcmI_filtered_final_distinct)[names(TcmI_filtered_final_distinct)=="cyclaseType"] <-"cyclase_type"

results_common<-TcmI_filtered_final_distinct %>% inner_join(., tcmI_blast_intervals, by = c("readID" = "sseqid", "Sample", "sampleType", "interval", "cyclase_type")) %>% select(-c(model_cov))
results_hmm_unique<-TcmI_filtered_final_distinct %>% anti_join(., tcmI_blast_intervals, by = c("readID" = "sseqid", "Sample", "sampleType", "interval", "cyclase_type")) 
results_hmm_unique_TP<-compare_hmm_unique(results_hmm_unique,all_t2pksBlastDF,tcmi_positions )

results_hmm_unique_FP <- results_hmm_unique %>% anti_join(.,results_hmm_unique_TP)


TcmI_filtered_final_distinct_TPs <- rbind(results_common, results_hmm_unique_TP) %>% select(-c(8:10,12:17))

combined_TcmI_results <- TcmI_filtered_final_distinct_TPs %>% full_join(., results_hmm_unique_FP )

#write_tsv(combined_TcmI_results,"TcmI-final_models-data.txt",col_names = T)
```


```{r}
sessionInfo()
```