---
title: "coverage_analysis-figure"
author: "Francine Camacho"
date: "7/20/2019"
output:
  html_document: 
    smart: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/francinecamacho/Google Drive/T2PKS-paper/computational_analysis/synthetic_genomes/revision-synthetic-genomes')
```


```{r, message = FALSE }
#load libraries
require(tidyverse)
require(ggpubr)
require(ggsci)
```

```{r, message = FALSE }

#load coverage data 
high <- read_csv("./metadata/high_complexity_cov.csv", col_names = T)
high$complexity<- "high"
high$cohort <- "synthetic_stool"
low <- read_csv("./metadata/low_complexity_cov.csv", col_names = T)
low$complexity<- "low"
low$cohort <- "synthetic_stool"

synthetic_data <- rbind(high, low) %>% unite(., "sample", c("cohort","Sample"), sep ="_", remove = T)
```

```{r}

#Download HMM data for each domains with final good models 
OxyN <- read_tsv("./HMM-analysis/OxyN/OxyN-final_data-raw.txt", col_names = T)
TcmN <- read_tsv("./HMM-analysis/TcmN/TcmN-final_data-raw.txt", col_names = T)
TcmI <- read_tsv("./HMM-analysis/TcmI/TcmI-final_data-raw.txt", col_names = T)
TcmJ <- read_tsv("./HMM-analysis/TcmJ/TcmJ-final_data-raw.txt", col_names = T)

hmm_df <- rbind(OxyN, TcmN, TcmI, TcmJ )
cyclase_hmm_counts <- hmm_df %>% group_by(Sample, cyclaseType) %>% tally() %>% ungroup()

# Test for TcmI; remove samples that do not include @ least one TcmI genome 
fig_per_cyclase <- function(hmm_counts, synthetic, cyclase_name, pks_genomes){
  
    synthetic_noPKS <- synthetic %>% filter(str_detect(genome, "/tigress/DONIA/fcamacho/TypeII-PKS-Manuscript/synthetic-genomes/background_genomes/"))%>% separate(., genome, c("a", "b"), sep ="/background_genomes/",remove =T) %>% mutate(genome = str_replace(b, ".fa","")) %>% select(-c(a,b))
  synthetic_PKS <- synthetic %>% filter(str_detect(genome,"/tigress/DONIA/fcamacho/TypeII-PKS-Manuscript/synthetic-genomes/T2PKS-Positive-genomes/T2PKS_pos-genomes-synthetic-metagenome/")) %>% separate(., genome, c("a", "b"), sep ="/T2PKS_pos-genomes-synthetic-metagenome/",remove =T) %>% mutate(genome = str_replace(b, ".fasta","")) %>% select(-c(a,b))
  synthetic_df <- rbind(synthetic_noPKS,synthetic_PKS)
  synthetic_df <- synthetic_df[c("sample", "complexity", "genome", "coverage")]
  synthetic_df <- synthetic_df %>% mutate(pks = ifelse( genome %in% pks_genomes, 1,0))
  #synthetic_counts <- synthetic_df %>% group_by(sample,pks) %>%  mutate(sumCov = ifelse(pks == 1 & !genome %in% "GCA", sum(coverage),0)) %>% ungroup()
    synthetic_counts <- synthetic_df %>% group_by(sample,pks) %>%  mutate(sumCov = ifelse(pks == 1, sum(coverage),0)) %>% ungroup()
  synthetic_coverage <- synthetic_counts %>% distinct(sample, complexity, sumCov) %>% group_by(sample) %>% arrange(desc(sumCov)) %>% top_n(1, sumCov) %>% ungroup()
  results_cyclase <- hmm_counts %>% inner_join(., synthetic_coverage, by = c("Sample"="sample")) %>% filter(cyclaseType == cyclase_name)
  plot_cyclase <- results_cyclase %>% ggplot(., aes(sumCov, n,  color = cyclaseType))+ geom_point() + facet_wrap(~complexity) + theme_pubclean() + scale_color_npg() + geom_smooth(aes(colour = cyclaseType), method = "loess", se = TRUE) + xlab("Genome Coverage")+ ylab("Total number of detected s-HMM reads")
  plot(plot_cyclase)
  return(results_cyclase)

  
}

tcmi_genomes <- c("Salinispora_arenicola_CNB527", "Streptomyces_olivaceus_strain_NRRL_B-3009", "Streptomyces_olindensis_strain_DAUFPE_5622", "Streptomyces_collinus_Tu_365")

results_tcmi <- fig_per_cyclase(cyclase_hmm_counts,synthetic_data, "TcmI", tcmi_genomes )

# check samples 
#synthetic_data %>% filter(sample == "synthetic_stool_high_S13") %>% filter(!str_detect(genome, "GCA"))
#TcmI_coverage-scatterplot , 8 vs 10 size demisons 

#check spearman correlations 
results_tcmi %>% filter(complexity == "high") %>% select(n,sumCov) %>% cor(., method = "spearman")
results_tcmi %>% filter(complexity == "low") %>% select(n,sumCov) %>% cor(., method = "spearman")

#############
tcmn_genomes <- c("Bacillus_endophyticus_DSM_13796", "Clostridium_beijerinckii_strain_DSM_53", "Orenia_metallireducens_strain_MSL47", "Paenibacillus_borealis_strain_DSM_13188", "Photorhabdus_luminescens_strain_ATCC_29999", "Salinispora_arenicola_CNB527", "Streptomyces_collinus_Tu_365", "Streptomyces_olindensis_strain_DAUFPE_5622", "Streptomyces_olivaceus_strain_NRRL_B-3009", "Thermoactinomyces_vulgaris_strain_NRRL_F-5595")
results_tcmn <- fig_per_cyclase(cyclase_hmm_counts,synthetic_data, "TcmN", tcmn_genomes )

results_tcmn %>% filter(complexity == "high") %>% select(n,sumCov) %>% cor(., method = "spearman")
results_tcmn %>% filter(complexity == "low") %>% select(n,sumCov) %>% cor(., method = "spearman")
##########
tcmj_genomes <- c("Bacillus_endophyticus_DSM_13796", "Orenia_metallireducens_strain_MSL47", "Paenibacillus_borealis_strain_DSM_13188", "Salinispora_arenicola_CNB527", "Streptomyces_collinus_Tu_365", "Streptomyces_olindensis_strain_DAUFPE_5622", "Streptomyces_olivaceus_strain_NRRL_B-3009", "Thermoactinomyces_vulgaris_strain_NRRL_F-5595")
results_tcmj <- fig_per_cyclase(cyclase_hmm_counts,synthetic_data, "TcmJ", tcmj_genomes )

results_tcmj %>% filter(complexity == "high") %>% select(n,sumCov) %>% cor(., method = "spearman")
results_tcmj %>% filter(complexity == "low") %>% select(n,sumCov) %>% cor(., method = "spearman")
##########
oxyn_genomes <-c("Bacillus_endophyticus_DSM_13796", "Clostridium_beijerinckii_strain_DSM_53", "Orenia_metallireducens_strain_MSL47", "Paenibacillus_borealis_strain_DSM_13188", "Photorhabdus_luminescens_strain_ATCC_29999", "Streptomyces_collinus_Tu_365", "Streptomyces_olindensis_strain_DAUFPE_5622", "Thermoactinomyces_vulgaris_strain_NRRL_F-5595")
results_oxyn <- fig_per_cyclase(cyclase_hmm_counts,synthetic_data, "OxyN", oxyn_genomes )

results_oxyn %>% filter(complexity == "high") %>% select(n,sumCov) %>% cor(., method = "spearman")
results_oxyn %>% filter(complexity == "low") %>% select(n,sumCov) %>% cor(., method = "spearman")
##########
```


```{r}
#This plots a linear model with ggplot2. It is a function written by Susan Johnston 
ggplotRegression <- function (fit) {
ggplot(fit$model, aes_string(x = names(fit$model)[2], y = names(fit$model)[1])) + 
  geom_point() + theme_pubclean() + scale_color_npg() + xlab("Genome Coverage")+  ylab("Total number of detected s-HMM reads")+
  stat_smooth(method = "lm", col = "red") +
  labs(title = paste("Adj R2 = ",signif(summary(fit)$adj.r.squared, 5),
                     "Intercept =",signif(fit$coef[[1]],5 ),
                     " Slope =",signif(fit$coef[[2]], 5),
                     " P =",signif(summary(fit)$coef[2,4], 5)))
}

ggplotRegression(lm(n ~ sumCov, data = results_tcmi %>% filter(complexity == "high")))
ggplotRegression(lm(n ~ sumCov, data = results_tcmi %>% filter(complexity == "low")))

ggplotRegression(lm(n ~ sumCov, data = results_tcmn %>% filter(complexity == "high")))
ggplotRegression(lm(n ~ sumCov, data = results_tcmn %>% filter(complexity == "low")))

ggplotRegression(lm(n ~ sumCov, data = results_tcmj %>% filter(complexity == "high")))
ggplotRegression(lm(n ~ sumCov, data = results_tcmj %>% filter(complexity == "low")))

ggplotRegression(lm(n ~ sumCov, data = results_oxyn %>% filter(complexity == "high")))
ggplotRegression(lm(n ~ sumCov, data = results_oxyn %>% filter(complexity == "low")))
```