---
title: "full-OxyN-analysis"
author: "Francine Camacho"
date: "7/20/2019"
output:
  html_document: 
    smart: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/francinecamacho/Google Drive/T2PKS-paper/computational_analysis/synthetic_genomes/revision-synthetic-genomes/HMM-analysis/OxyN/')
```


```{r, message = FALSE }
#load libraries
require(tidyverse)
```

### Load full profiled cyclase OxyN HMMs for synthetic genomes
```{r, message = FALSE}
#load OxyN HMM data and recode sampleType for the complexity of the synthetic sample (#of genomes in samples)
cyclase_OxyN <- read_delim("combined-cyclase_OxyN-full-synthetic_genomes-results.txt",col_names = F, delim = "\t") 
names(cyclase_OxyN) <- c("readID", "sampleType", "sampleID", "cyclaseType", "HMMScore")
```

##### Remove the reading frame information from the readID. 
```{r,message=FALSE}
# Function to aggregate identitical sample reads located at different reading frames 
# and take the frame with the highest HMM score
formatHMM<-function(hmmdf){
  hmmdfRecoded <- separate(hmmdf, readID, into = c("readIDOnly","F_R_read_frame"), sep = "/", extra = "merge")
  hmmdfRecoded_FR <- separate(hmmdfRecoded, F_R_read_frame, into = c("F_R","frameNumb"), sep = "_", extra = "merge")
  hmmdfRecodedDF<- within(hmmdfRecoded_FR, readID <- paste(readIDOnly,F_R, sep='/'))
  hmmdfRecodedDFUnique<-aggregate(HMMScore ~ readID + sampleID + sampleType + cyclaseType , hmmdfRecodedDF, max)
  colnames(hmmdfRecodedDFUnique)<-c("readID","Sample", "sampleType", "cyclaseType","HMMScore")
  return(hmmdfRecodedDFUnique)
}

#Keep duplicated reads if they are in different reads
cyclaseOxyN_recoded<-formatHMM(cyclase_OxyN)
```

##### Load the BLAST data for T2PKS cyclase genes against our synthetic dataset. 
```{r, message=FALSE}
#BLAST True positive reads mapped to the T2PKS cyclases genes 
cyclase_metadata <- read_tsv("../../../HMM-analysis/T2PKS-pos_genomes-cyclase_gene-metadata.txt", col_names = T) %>% filter(used_in_synthetic_metagenome == "Y") %>% select(c(11:9))

# BLAST filtered data 
t2pksBlastDF <- read_delim("../../T2PKS-TP-reads/t2pksBlastDF_filter-results.txt", col_names = T, delim = "\t") %>% inner_join(.,cyclase_metadata, by= c("qseqid" = "nucl_cyclase_target_name") )%>% filter(cyclase_type == "OxyN") 

```

##### Determine manual cutoffs for each interval within the model
```{r, message=FALSE}
compareCyclaseReads<-function(cyclaseDF, blastDF){
  
  names(blastDF)[names(blastDF)=="sseqid"] <-"readID"
  names(blastDF)[names(blastDF)=="qseqid"] <-"bgcName"
  names(cyclaseDF)[names(cyclaseDF)=="cyclaseType"] <-"cyclase_type"
  
  common_reads <- blastDF %>% inner_join(.,cyclaseDF) %>% select(-c(HMMScore))
  common_reads$readCheck<-"common-read"
  
  hmm_unique_reads <- cyclaseDF %>% anti_join(.,blastDF)
  hmm_unique_reads$readCheck<-"hmm-unique-read"
  blast_unique_reads <- blastDF %>% anti_join(.,cyclaseDF)
  blast_unique_reads$readCheck<-"blast-unique-read"
  
  blast_common_data <- rbind(common_reads, blast_unique_reads)
  allCyclases_TP<- blast_common_data %>% full_join(.,hmm_unique_reads) 
  allCyclases_TP$bgcName[is.na(allCyclases_TP$bgcName)]<-"hmm-unique-read"
  return(allCyclases_TP %>% select(-c(HMMScore)))
  
}
##############################################################################################
cyclaseOxyN_bin <- compareCyclaseReads(cyclaseOxyN_recoded,t2pksBlastDF)
names(cyclaseOxyN_bin)[names(cyclaseOxyN_bin)=="cyclase_type"] <-"cyclaseType"
```

##### Trying to determine the HMM cutoffs; calculate median 

```{r, message=FALSE}

compare_reads <- function(hmm_df, blast_df){
  names(blast_df)[names(blast_df)=="sseqid"] <-"readID"
  names(blast_df)[names(blast_df)=="qseqid"] <-"bgcName"
  names(hmm_df)[names(hmm_df)=="cyclaseType"] <-"cyclase_type"
  
  common_reads <- hmm_df %>% semi_join(.,blast_df) 
  common_reads$readCheck<-"common-read"
  hmm_unique_reads <- hmm_df %>% anti_join(.,blast_df)
  hmm_unique_reads$readCheck<-"hmm-unique-read"
  compared_data <- rbind(common_reads, hmm_unique_reads)
  return(compared_data)
}

compare_reads(cyclaseOxyN_recoded, t2pksBlastDF) %>% filter(readCheck == "common-read") %>% mutate(medianScore = round(median(HMMScore))) %>% distinct(readCheck, medianScore) %>% ungroup()

```

### Apply different HMM Scores cutoff to the HMM dataset
```{r,message=FALSE}
#Filter OxyN data with cutoffs to compare to BLAST interval reads 
OxyN_filtered_median <- cyclaseOxyN_recoded %>% filter(HMMScore >=35) %>% inner_join(.,cyclaseOxyN_bin)

#OxyN_filtered_median  %>% filter(readCheck == "hmm-unique-read") %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% distinct(scaffold_id)

pks_genomes_median_check <-c("gi|928420278|gb|LGKI01000101.1|", "gi|1081600317|emb|FMWJ01000012.1|", "gi|928414697|gb|LGKI01000478.1|", 
                             "LZZI01000003.1", "gi|529222834|ref|NC_021985.1|", 
                             "OBDZ01000030.1")

remaining_hmm_OxyN_pos_median_check <- OxyN_filtered_median %>% filter(readCheck == "hmm-unique-read") %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% filter(scaffold_id %in% pks_genomes_median_check)

####################################################################################
OxyN_filtered_median_subfive<- cyclaseOxyN_recoded %>% filter(HMMScore >=30) %>% inner_join(.,cyclaseOxyN_bin)

#OxyN_filtered_median_subfive  %>% filter(readCheck == "hmm-unique-read") %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% distinct(scaffold_id)

pks_genomes_subfive_check <-c("gi|928416322|gb|LGKI01000367.1|", "gi|928420278|gb|LGKI01000101.1|", "gi|1098137419|emb|FOXX01000002.1|", "gi|1081600317|emb|FMWJ01000012.1|", "gi|640937269|gb|JJOH01000002.1|", "gi|928414697|gb|LGKI01000478.1|", "LZZI01000003.1", "OBDZ01000030.1", "gi|529222834|ref|NC_021985.1|")
remaining_hmm_OxyN_pos_subfive_check <- OxyN_filtered_median_subfive %>% filter(readCheck == "hmm-unique-read") %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% filter(scaffold_id %in% pks_genomes_subfive_check)
####################################################################################                                 
OxyN_filtered_median_plusfive <- cyclaseOxyN_recoded %>% filter(HMMScore>=40) %>% inner_join(.,cyclaseOxyN_bin)

#OxyN_filtered_median_plusfive  %>% filter(readCheck == "hmm-unique-read") %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% distinct(scaffold_id)
pks_genomes_plusfive_check <- c("gi|928420278|gb|LGKI01000101.1|", "gi|1081600317|emb|FMWJ01000012.1|", "gi|928414697|gb|LGKI01000478.1|", "LZZI01000003.1")
remaining_hmm_OxyN_pos_plusfive_check <- OxyN_filtered_median_plusfive %>% filter(readCheck == "hmm-unique-read") %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% filter(scaffold_id %in% pks_genomes_plusfive_check)

```


```{r}
#Function to parse out HMM unique reads for each sample. 
parseReads <- function(HMMdf, modelName, dirname){
  setwd(dirname)
  samples<-unique(HMMdf$Sample)
  for (s in 1:length(samples)){
    currentSample<-samples[s]
    currentSampleResults<- HMMdf %>% filter(Sample == currentSample)
    currentSampleReads<- unique(currentSampleResults$readID)
    fileName<-paste0(currentSample,paste("-",modelName,"-hmm_unique",sep =""), ".txt", sep ="")
    write.table(currentSampleReads,fileName, quote = F, row.names = F, col.names = F )
  }
}

#parseReads(remaining_hmm_OxyN_pos_median_check, "cyclaseOxyN", "/Users/francinecamacho/Google Drive/T2PKS-paper/computational_analysis/synthetic_genomes/revision-synthetic-genomes/HMM-analysis/OxyN/full_model-analysis/hmm-unique-analysis/median_score/")

#parseReads(remaining_hmm_OxyN_pos_subfive_check, "cyclaseOxyN", "/Users/francinecamacho/Google Drive/T2PKS-paper/computational_analysis/synthetic_genomes/revision-synthetic-genomes/HMM-analysis/OxyN/full_model-analysis/hmm-unique-analysis/minus_five_score")

#parseReads(remaining_hmm_OxyN_pos_plusfive_check, "cyclaseOxyN", "/Users/francinecamacho/Google Drive/T2PKS-paper/computational_analysis/synthetic_genomes/revision-synthetic-genomes/HMM-analysis/OxyN/full_model-analysis/hmm-unique-analysis/plus_five_score/")

```

```{r, message=FALSE}

calculate_F1 <- function(df){
  TP <- df %>% filter(readCheck == "common-read") %>% nrow()
  FP <- df %>% filter(readCheck == "hmm-unique-read") %>% nrow()
  FN <- df %>% filter(readCheck == "blast-unique-read") %>% nrow()
  F1_metric <- (2*TP)/((2*TP)+FP+FN)
  print(F1_metric)
  return(F1_metric)
  
  
}
  
remove_OxyN_reads_median <- data.frame(
  readID = c("gi|1081600317|emb|FMWJ01000012.1|-102984/2",  "gi|928414697|gb|LGKI01000478.1|-10802/2"), 
  Sample = c("synthetic_stool_low_S11", "synthetic_stool_low_S22")) 

remove_OxyN_reads_plusfive <- data.frame(
  readID = c("gi|1081600317|emb|FMWJ01000012.1|-102984/2",  "gi|928414697|gb|LGKI01000478.1|-10802/2"), 
  Sample = c("synthetic_stool_low_S11", "synthetic_stool_low_S22")) 

remove_OxyN_reads_subfive <- data.frame(
  readID = c("gi|1081600317|emb|FMWJ01000012.1|-45808/2", "gi|640937269|gb|JJOH01000002.1|-41502/2", "gi|1081600317|emb|FMWJ01000012.1|-15560/2", "gi|1081600317|emb|FMWJ01000012.1|-364990/2", "gi|1081600317|emb|FMWJ01000012.1|-260274/2", "gi|1081600317|emb|FMWJ01000012.1|-138300/2",
  "gi|1081600317|emb|FMWJ01000012.1|-102984/2", "gi|1081600317|emb|FMWJ01000012.1|-62068/1",
  "gi|1081600317|emb|FMWJ01000012.1|-862/1", "gi|928414697|gb|LGKI01000478.1|-10802/2", 
  "gi|1081600317|emb|FMWJ01000012.1|-10958/2", "gi|529222834|ref|NC_021985.1|-5158572/1", 
  "gi|529222834|ref|NC_021985.1|-3409650/1", "gi|529222834|ref|NC_021985.1|-1522794/1",
  "gi|529222834|ref|NC_021985.1|-6434954/1", "gi|529222834|ref|NC_021985.1|-1003728/1", 
  "gi|529222834|ref|NC_021985.1|-444108/2", "gi|529222834|ref|NC_021985.1|-748426/1",
  "gi|529222834|ref|NC_021985.1|-488776/2", "gi|529222834|ref|NC_021985.1|-7558568/1", 
  "gi|529222834|ref|NC_021985.1|-5701540/1", "gi|529222834|ref|NC_021985.1|-5056216/1", 
  "gi|529222834|ref|NC_021985.1|-4392974/1", "gi|529222834|ref|NC_021985.1|-2293396/2",
  "gi|640937269|gb|JJOH01000002.1|-14820/2", "gi|640937269|gb|JJOH01000002.1|-104902/2",
  "gi|640937269|gb|JJOH01000002.1|-86874/1", "gi|640937269|gb|JJOH01000002.1|-54572/2", 
  "gi|640937269|gb|JJOH01000002.1|-52570/1", "gi|640937269|gb|JJOH01000002.1|-6304/1", 
  "gi|640937269|gb|JJOH01000002.1|-43884/1"), 
  Sample = c("synthetic_stool_high_S24","synthetic_stool_high_S53", "synthetic_stool_low_S10","synthetic_stool_low_S11",
  "synthetic_stool_low_S11","synthetic_stool_low_S11","synthetic_stool_low_S11", "synthetic_stool_low_S11",
  "synthetic_stool_low_S16", "synthetic_stool_low_S22", "synthetic_stool_low_S29", "synthetic_stool_low_S32",
  "synthetic_stool_low_S32","synthetic_stool_low_S32", "synthetic_stool_low_S34", "synthetic_stool_low_S34",
  "synthetic_stool_low_S34", "synthetic_stool_low_S41", "synthetic_stool_low_S41", "synthetic_stool_low_S51",
  "synthetic_stool_low_S51", "synthetic_stool_low_S51", "synthetic_stool_low_S51", "synthetic_stool_low_S51",
  "synthetic_stool_low_S56", "synthetic_stool_low_S63", "synthetic_stool_low_S63", "synthetic_stool_low_S63",
  "synthetic_stool_low_S63", "synthetic_stool_low_S63", "synthetic_stool_low_S68")) 


OxyN_f1_cutoff_median <- calculate_F1(compareCyclaseReads(cyclaseOxyN_recoded %>% anti_join(., remove_OxyN_reads_median) %>% filter(HMMScore >=35),t2pksBlastDF)) #0.4887247
OxyN_f1_cutoff_plusfive <- calculate_F1(compareCyclaseReads(cyclaseOxyN_recoded %>% anti_join(., remove_OxyN_reads_plusfive) %>% filter(HMMScore >=40),t2pksBlastDF)) #0.3224454
OxyN_f1_cutoff_subfive <- calculate_F1(compareCyclaseReads(cyclaseOxyN_recoded %>% anti_join(., remove_OxyN_reads_subfive) %>% filter(HMMScore >=30),t2pksBlastDF)) #0.4995103
```

```{r}
sessionInfo()
```


