---
title: "full-lantibiotics-analysis"
author: "Francine Camacho"
date: "7/31/2019"
output:
  html_document: 
    smart: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = './synthetic_genomes/lantibiotics')
```


```{r, message = FALSE }
#load libraries
require(tidyverse)
```

### Load full profiled cyclase lantibiotics HMMs for synthetic genomes
```{r, message = FALSE}
#load lantibiotics HMM data and recode sampleType for the complexity of the synthetic sample (#of genomes in samples)
lantibiotics <- read_delim("combined-lantibiotics-full-synthetic_genomes-results.txt",col_names = F, delim = "\t") 
names(lantibiotics) <- c("readID", "sampleType", "sampleID", "cyclaseType", "HMMScore")
```

# remove Tp+other genomes that antiSMASH failed to identify 
```{r}
# We need to remove reads that are from TP+Other genomes that were missed by antiSMASH 
remove_reads <- read_tsv("blastx-fp-subfive.txt", col_names =  T) %>% filter(readCheck == "TP" & lanc_genome == "other") %>% select(c(readID, Sample))
```

##### Remove the reading frame information from the readID. 
```{r,message=FALSE}
# Function to aggregate identitical sample reads located at different reading frames 
# and take the frame with the highest HMM score
formatHMM<-function(hmmdf){
  hmmdfRecoded <- separate(hmmdf, readID, into = c("readIDOnly","F_R_read_frame"), sep = "/", extra = "merge")
  hmmdfRecoded_FR <- separate(hmmdfRecoded, F_R_read_frame, into = c("F_R","frameNumb"), sep = "_", extra = "merge")
  hmmdfRecodedDF<- within(hmmdfRecoded_FR, readID <- paste(readIDOnly,F_R, sep='/'))
  hmmdfRecodedDFUnique<-aggregate(HMMScore ~ readID + sampleID + sampleType + cyclaseType , hmmdfRecodedDF, max)
  colnames(hmmdfRecodedDFUnique)<-c("readID","Sample", "sampleType", "cyclaseType","HMMScore")
  return(hmmdfRecodedDFUnique)
}

#Keep duplicated reads if they are in different reads
lantibiotics_recoded<-formatHMM(lantibiotics)%>% anti_join(.,remove_reads)
```

##### Load the BLAST data for lantibiotics genes against our synthetic dataset. 
```{r, message=FALSE}
# BLAST unfiltered reads at 95% pident no readCoverage filter
all_lancBlastDF <- read_delim("../Lantibiotics-TP-reads/lantibioticBlastDF_filter-results-updated.txt", col_names = T, delim = "\t")%>% anti_join(.,remove_reads, by= c("sseqid"= "readID", "Sample"))
all_lancBlastDF$cyclase_type <-"lantibiotic"
```

##### Determine manual cutoffs for each interval within the model
```{r, message=FALSE}
compareCyclaseReads<-function(cyclaseDF, blastDF){
  
  names(blastDF)[names(blastDF)=="sseqid"] <-"readID"
  names(blastDF)[names(blastDF)=="qseqid"] <-"bgcName"
  names(cyclaseDF)[names(cyclaseDF)=="cyclaseType"] <-"cyclase_type"
  
  common_reads <- blastDF %>% inner_join(.,cyclaseDF) %>% select(-c(HMMScore))
  common_reads$readCheck<-"common-read"
  
  hmm_unique_reads <- cyclaseDF %>% anti_join(.,blastDF)
  blast_unique_reads <- blastDF %>% anti_join(.,cyclaseDF)
  blast_unique_reads$readCheck<-"blast-unique-read"
  if (nrow(hmm_unique_reads) >0){
    hmm_unique_reads$readCheck<-"hmm-unique-read"
    blast_common_data <- rbind(common_reads, blast_unique_reads)
    allCyclases_TP<- blast_common_data %>% full_join(.,hmm_unique_reads) 
    allCyclases_TP$bgcName[is.na(allCyclases_TP$bgcName)]<-"hmm-unique-read"
    return(allCyclases_TP %>% select(-c(HMMScore)))
    
  }else{
    blast_common_data <- rbind(common_reads, blast_unique_reads)
    return(blast_common_data)
  }
}
##############################################################################################
lanc_bin <- compareCyclaseReads(lantibiotics_recoded,all_lancBlastDF)
names(lanc_bin)[names(lanc_bin)=="cyclase_type"] <-"cyclaseType"
```

##### Trying to determine the HMM cutoffs; calculate median 

```{r, message=FALSE}

compare_reads <- function(hmm_df, blast_df){
  names(blast_df)[names(blast_df)=="sseqid"] <-"readID"
  names(blast_df)[names(blast_df)=="qseqid"] <-"bgcName"
  names(hmm_df)[names(hmm_df)=="cyclaseType"] <-"cyclase_type"
  
  common_reads <- hmm_df %>% semi_join(.,blast_df) 
  common_reads$readCheck<-"common-read"
  hmm_unique_reads <- hmm_df %>% anti_join(.,blast_df)
  hmm_unique_reads$readCheck<-"hmm-unique-read"
  compared_data <- rbind(common_reads, hmm_unique_reads)
  return(compared_data)
}

compare_reads(lantibiotics_recoded, all_lancBlastDF) %>% filter(readCheck == "common-read") %>% mutate(medianScore = round(median(HMMScore))) %>% distinct(readCheck, medianScore) %>% ungroup()

```

### Apply different HMM Scores cutoff to the HMM dataset
```{r,message=FALSE}
#Filter lantibiotics data with cutoffs to compare to BLAST interval reads 
lanc_filtered_median <- lantibiotics_recoded %>% filter(HMMScore >=25) %>% inner_join(.,lanc_bin)

#lanc_filtered_median  %>% filter(readCheck == "hmm-unique-read") %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% distinct(scaffold_id)

lanc_genomes_median_check <-c("CP001834.1")

remaining_hmm_lanc_pos_median_check <- lanc_filtered_median %>% filter(readCheck == "hmm-unique-read") %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% filter(scaffold_id %in% lanc_genomes_median_check)

####################################################################################
lanc_filtered_median_subfive<- lantibiotics_recoded %>% filter(HMMScore >=20) %>% inner_join(.,lanc_bin)

#lanc_filtered_median_subfive  %>% filter(readCheck == "hmm-unique-read") %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% distinct(scaffold_id)

lanc_genomes_subfive_check <-c("HE613569.1", "CP001834.1", "CP002365.1")


remaining_hmm_lanc_pos_subfive_check <- lanc_filtered_median_subfive %>% filter(readCheck == "hmm-unique-read") %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% filter(scaffold_id %in% lanc_genomes_subfive_check)
####################################################################################                                 
lanc_filtered_median_plusfive <- lantibiotics_recoded %>% filter(HMMScore>=30) %>% inner_join(.,lanc_bin)

#lanc_filtered_median_plusfive  %>% filter(readCheck == "hmm-unique-read") %>% separate(readID, c("scaffold_id", "rest_id"), sep = "-", remove = F) %>% distinct(scaffold_id)
remaining_hmm_lanc_pos_plusfive_check <- c()
```


```{r}
#Function to parse out HMM unique reads for each sample. 
parseReads <- function(HMMdf, modelName, dirname){
  setwd(dirname)
  samples<-unique(HMMdf$Sample)
  for (s in 1:length(samples)){
    currentSample<-samples[s]
    currentSampleResults<- HMMdf %>% filter(Sample == currentSample)
    currentSampleReads<- unique(currentSampleResults$readID)
    fileName<-paste0(currentSample,paste("-",modelName,"-hmm_unique",sep =""), ".txt", sep ="")
    write.table(currentSampleReads,fileName, quote = F, row.names = F, col.names = F )
  }
}

parseReads(remaining_hmm_lanc_pos_median_check, "lantibiotics", "/Users/francinecamacho/Google Drive/T2PKS-paper/computational_analysis/synthetic_genomes/revision-synthetic-genomes/other-BGCs-poc/lantibiotics/HMM-analysis/full_model-analysis/hmm-unique-analysis/median_score/")

parseReads(remaining_hmm_lanc_pos_subfive_check, "lantibiotics", "/Users/francinecamacho/Google Drive/T2PKS-paper/computational_analysis/synthetic_genomes/revision-synthetic-genomes/other-BGCs-poc/lantibiotics/HMM-analysis/full_model-analysis/hmm-unique-analysis/minus_five_score")

#No plusfive read scores that are belonging to the lanc positive genomes 

```

```{r, message=FALSE}
#Load hmm-unique blast to sid domains reads to remove them from the F1 analysis 
lanc_hmmunique_blast <- read_tsv("full_model-analysis/hmm-unique-analysis/lanc_hmm-unique_blast-results.txt", col_names = T)%>% separate(qseqid, c("scaffold_id", "rest_id"), sep = "-", remove = F)%>%
  filter(str_detect(sseqid, fixed(scaffold_id))) %>% select(-c(scaffold_id, rest_id))  %>% arrange(evalue) %>% group_by(qseqid, Sample, cutoff) %>% top_n(1) %>% ungroup()

calculate_F1 <- function(df){
  TP <- df %>% filter(readCheck == "common-read") %>% nrow()
  FP <- df %>% filter(readCheck == "hmm-unique-read") %>% nrow()
  FN <- df %>% filter(readCheck == "blast-unique-read") %>% nrow()
  F1_metric <- (2*TP)/((2*TP)+FP+FN)
  print(F1_metric)
  return(F1_metric)
  
  
}

remove_lanc_reads_subfive <- lanc_hmmunique_blast %>% filter(cutoff == "minus_five_score") %>% select(qseqid, Sample) %>% distinct()
names(remove_lanc_reads_subfive)[1] <- "readID"

remove_lanc_reads_median <- lanc_hmmunique_blast %>% filter(cutoff == "median_score") %>% select(qseqid, Sample) %>% distinct()
names(remove_lanc_reads_median)[1] <- "readID"
remove_lanc_reads_plusfive <- lanc_hmmunique_blast %>% filter(cutoff == "plus_five_score") %>% select(qseqid, Sample) %>% distinct()
names(remove_lanc_reads_plusfive)[1] <- "readID"

lanc_f1_cutoff_median <-  calculate_F1(compareCyclaseReads(lantibiotics_recoded%>% anti_join(., remove_lanc_reads_median) %>% filter(HMMScore >=25),all_lancBlastDF)) #0.04341304
lanc_f1_cutoff_plusfive <- calculate_F1(compareCyclaseReads(lantibiotics_recoded %>% anti_join(., remove_lanc_reads_plusfive) %>% filter(HMMScore >=30),all_lancBlastDF)) #0.003587805
lanc_f1_cutoff_subfive <- calculate_F1(compareCyclaseReads(lantibiotics_recoded %>% anti_join(., remove_lanc_reads_subfive) %>% filter(HMMScore >=20),all_lancBlastDF)) #0.1051712
```
##The highest cutoff for the full HMM is with cutoff of 20 and has an F1 of 10.5.
```{r}
sessionInfo()
```


